File: webp-imageio/src/test/java/com/luciad/imageio/webp/WebPTest.java
package com.luciad.imageio.webp;

import org.junit.jupiter.api.Test;

import javax.imageio.*;
import javax.imageio.stream.MemoryCacheImageInputStream;
import javax.imageio.stream.MemoryCacheImageOutputStream;
import java.awt.image.BufferedImage;
import java.awt.image.DataBufferInt;
import java.io.*;
import java.util.Iterator;
import java.util.Random;

import static org.junit.jupiter.api.Assertions.*;

public class WebPTest {

    @Test
    public void testFindReaderByMimeType() {
        assertNotNull(
                findReader(ImageIO.getImageReadersByMIMEType("image/webp")),
                "Image reader was not registered"
        );
    }

    @Test
    public void testFindReaderByFormatName() {
        assertNotNull(
                findReader(ImageIO.getImageReadersByFormatName("webp")),
                "Image reader was not registered"
        );
    }

    @Test
    public void testFindReaderBySuffix() {
        assertNotNull(
                findReader(ImageIO.getImageReadersBySuffix("webp")),
                "Image reader was not registered"
        );
    }

    @Test
    public void testFindWriterByMimeType() {
        assertNotNull(
                findWriter(ImageIO.getImageWritersByMIMEType("image/webp")),
                "Image writer was not registered"
        );
    }

    @Test
    public void testFindWriterByFormatName() {
        assertNotNull(
                findWriter(ImageIO.getImageWritersByFormatName("webp")),
                "Image writer was not registered"
        );
    }

    @Test
    public void testFindWriterBySuffix() {
        assertNotNull(
                findWriter(ImageIO.getImageWritersBySuffix("webp")),
                "Image writer was not registered"
        );
    }

    @Test
    public void testDecompressLossy() throws IOException {
        byte[] webpData = readResource("lossy.webp");
        BufferedImage image = decompress(webpData);
        assertEquals(1024, image.getWidth());
        assertEquals(752, image.getHeight());
    }

    @Test
    public void testDecompressLossless() throws IOException {
        byte[] webpData = readResource("lossless.webp");
        BufferedImage image = decompress(webpData);
        assertEquals(400, image.getWidth());
        assertEquals(301, image.getHeight());
    }

    @Test
    public void testDecompressLossyAlpha() throws IOException {
        byte[] webpData = readResource("lossy_alpha.webp");
        BufferedImage image = decompress(webpData);
        assertEquals(400, image.getWidth());
        assertEquals(301, image.getHeight());
    }

    private byte[] readResource(String resource) throws IOException {
        InputStream stream = getClass().getClassLoader().getResourceAsStream(resource);
        if (stream == null) {
            throw new FileNotFoundException("Could not load resource " + resource);
        }

        try {
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            try {
                byte[] buffer = new byte[4096];
                int bytesRead;
                while ((bytesRead = stream.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            } finally {
                out.close();
            }
            return out.toByteArray();
        } finally {
            stream.close();
        }
    }

    private BufferedImage decompress(byte[] webp) throws IOException {
        ImageReader reader = getImageReader(webp);
        assertNotNull(reader);

        reader.setInput(new MemoryCacheImageInputStream(new ByteArrayInputStream(webp)));
        return reader.read(0);
    }

    @Test
    public void testCompress() throws IOException {
        BufferedImage image = new BufferedImage(256, 256, BufferedImage.TYPE_INT_ARGB);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        MemoryCacheImageOutputStream imageOut = new MemoryCacheImageOutputStream(out);

        ImageWriter writer = getImageWriter();
        writer.setOutput(imageOut);
        writer.write(image);

        imageOut.close();
        out.close();

        byte[] data = out.toByteArray();
        assertNotEquals(0, data.length);
        assertEquals('R', data[0] & 0xFF);
        assertEquals('I', data[1] & 0xFF);
        assertEquals('F', data[2] & 0xFF);
        assertEquals('F', data[3] & 0xFF);
    }

    @Test
    public void testRoundtrip() throws IOException {
        BufferedImage image = new BufferedImage(256, 256, BufferedImage.TYPE_INT_ARGB);
        Random rng = new Random(42);
        int[] buffer = ((DataBufferInt) image.getData().getDataBuffer()).getData();
        for (int i = 0; i < buffer.length; i++) {
            buffer[i] = rng.nextInt();
        }

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        MemoryCacheImageOutputStream imageOut = new MemoryCacheImageOutputStream(out);

        ImageWriter writer = getImageWriter();
        writer.setOutput(imageOut);
        WebPWriteParam writeParam = (WebPWriteParam) writer.getDefaultWriteParam();
        writeParam.setCompressionMode(ImageWriteParam.MODE_EXPLICIT);
        writeParam.setCompressionType("Lossless");
        writer.write(null, new IIOImage(image, null, null), writeParam);

        imageOut.close();
        out.close();

        byte[] webpData = out.toByteArray();

        ImageReader reader = getImageReader();
        reader.setInput(new MemoryCacheImageInputStream(new ByteArrayInputStream(webpData)));
        BufferedImage decodedImage = reader.read(0);

        for (int x = 0; x < image.getWidth(); x++) {
            for (int y = 0; y < image.getHeight(); y++) {
                assertEquals(image.getRGB(x, y), decodedImage.getRGB(x, y));
            }
        }
    }

    private ImageWriter getImageWriter() {
        return findWriter(ImageIO.getImageWritersByMIMEType("image/webp"));
    }

    private ImageWriter findWriter(Iterator<ImageWriter> writers) {
        ImageWriter writer = null;
        while (writers.hasNext()) {
            ImageWriter writerCandidate = writers.next();
            if (writerCandidate.getOriginatingProvider() instanceof WebPImageWriterSpi) {
                writer = writerCandidate;
                break;
            }
        }
        return writer;
    }

    private ImageReader getImageReader() {
        return findReader(ImageIO.getImageReadersByMIMEType("image/webp"));
    }

    private ImageReader getImageReader(byte[] data) {
        MemoryCacheImageInputStream stream = new MemoryCacheImageInputStream(new ByteArrayInputStream(data));
        return findReader(ImageIO.getImageReaders(stream));
    }

    private ImageReader findReader(Iterator<ImageReader> readers) {
        ImageReader reader = null;
        while (readers.hasNext()) {
            ImageReader readerCandidate = readers.next();
            if (readerCandidate.getOriginatingProvider() instanceof WebPImageReaderSpi) {
                reader = readerCandidate;
                break;
            }
        }
        return reader;
    }
}

File: webp-imageio/src/test/kotlin/com/luciad/imageio/webp/WebPTest.kt
package com.luciad.imageio.webp

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.awt.image.BufferedImage
import java.awt.image.DataBufferInt
import java.io.ByteArrayOutputStream
import java.nio.file.Path
import java.util.Random
import javax.imageio.ImageIO
import javax.imageio.stream.MemoryCacheImageOutputStream

class WebPTest {

    @Test
    fun testFindReaderByMimeType() {
        val readers = ImageIO.getImageReadersByMIMEType("image/webp")

        assertThat(readers.requireWebpImageReader()).isNotNull()
    }

    @Test
    fun testFindReaderByFormatName() {
        val readers = ImageIO.getImageReadersByFormatName("webp")

        assertThat(readers.requireWebpImageReader()).isNotNull()
    }

    @Test
    fun testFindReaderBySuffix() {
        val readers = ImageIO.getImageReadersBySuffix("webp")

        assertThat(readers.requireWebpImageReader()).isNotNull()
    }

    @Test
    fun testFindWriterByMimeType() {
        val writer = ImageIO.getImageWritersByMIMEType("image/webp")

        assertThat(writer.requireWebpImageWriter()).isNotNull()
    }

    @Test
    fun testFindWriterByFormatName() {
        val writer = ImageIO.getImageWritersByFormatName("webp")
        assertThat(writer.requireWebpImageWriter()).isNotNull()
    }

    @Test
    fun testFindWriterBySuffix() {
        val writer = ImageIO.getImageWritersBySuffix("webp")
        assertThat(writer.requireWebpImageWriter()).isNotNull()
    }

    @Test
    fun readLossy() {
        val image = readImage(readResource("lossy.webp"))

        assertThat(image.width).isEqualTo(1024)
        assertThat(image.height).isEqualTo(752)
    }

    @Test
    fun readLossless() {
        val image = readImage(readResource("lossless.webp"))

        assertThat(image.width).isEqualTo(400)
        assertThat(image.height).isEqualTo(301)
    }

    @Test
    fun readLossyWithAlpha() {
        val image = readImage(readResource("lossy_alpha.webp"))

        assertThat(image.width).isEqualTo(400)
        assertThat(image.height).isEqualTo(301)
    }

    @Test
    fun testCompress() {
        val image = BufferedImage(256, 256, BufferedImage.TYPE_INT_ARGB)
        val data = ByteArrayOutputStream().use { out ->
            MemoryCacheImageOutputStream(out).use { imageOut ->
                writeWebpImage(
                    input = image,
                    target = imageOut,
                )
            }
            out.toByteArray()
        }

        assertThat(data.size).isNotEqualTo(0)
        assertThat(data[0].toInt() and 0xFF).isEqualTo('R'.code)
        assertThat(data[1].toInt() and 0xFF).isEqualTo('I'.code)
        assertThat(data[2].toInt() and 0xFF).isEqualTo('F'.code)
        assertThat(data[3].toInt() and 0xFF).isEqualTo('F'.code)
    }

    @Test
    fun testRoundTrip() {
        val image = BufferedImage(256, 256, BufferedImage.TYPE_INT_ARGB)
        val rng = Random(42)
        val buffer = (image.data.dataBuffer as DataBufferInt).data
        for (i in buffer.indices) {
            buffer[i] = rng.nextInt()
        }
        val out = ByteArrayOutputStream()
        val imageOut = MemoryCacheImageOutputStream(out)

        writeWebpImage(
            input = image,
            target = imageOut,
            params = {
                compressionType = CompressionType.Lossless
            },
        )
        imageOut.close()
        out.close()
        val webpData = out.toByteArray()

        val decodedImage = readImage(webpData)
        for (x in 0 until image.width) {
            for (y in 0 until image.height) {
                assertThat(image.getRGB(x, y)).isEqualTo(decodedImage.getRGB(x, y))
            }
        }
    }

    @Test
    fun canUseAllReadOptions() {
        val readParam = WebPReadParam().apply {
            bypassFiltering = true
            noFancyUpsampling = true
            useCropping = true
            cropLeft = 1
            cropTop = 1
            cropWidth = 380
            cropHeight = 260
            useScaling = true
            scaledWidth = 200
            scaledHeight = 150
            useThreads = true
            ditheringStrength = 20
            flipVertically = true
            alphaDitheringStrength = 1
        }
        val output = readImage(
            webp = readResource("lossless.webp"),
            param = readParam,
        )

        assertThat(output).isNotNull()
        assertThat(output.width).isEqualTo(200)
        assertThat(output.height).isEqualTo(150)
    }

    @Test
    fun canUseAllWriteOptions(@TempDir tempDir: Path) {
        val inputImage = ImageIO.read(getResourceStream("test4.png"))
        val outputFile = tempDir.resolve("output.webp").toFile()

        writeWebpImage(
            input = inputImage,
            target = outputFile,
            params = {
                compressionQuality = 1f
                compressionType = CompressionType.Lossy
                method = 6
                targetSize = 0
                targetPSNR = 0f
                segments = 4
                snsStrength = 2
                filterStrength = 2
                filterSharpness = 7
                filterType = 1
                autoAdjustFilterStrength = false
                alphaCompressionAlgorithm = 1
                alphaFiltering = 1
                alphaQuality = 100
                entropyAnalysisPassCount = 7
                showCompressed = false
                preprocessing = 0
                partitions = 0
                partitionLimit = 0
                emulateJpegSize = false
                threadLevel = 0
                lowMemory = false
                nearLossless = 100
                exact = false
                useDeltaPalette = false
                useSharpYUV = true
                qMin = 0
                qMax = 0
            },
        )
    }

    @Test
    fun sharpYuv(@TempDir tempDir: Path) {
        val inputImage = ImageIO.read(getResourceStream("test4.png"))
        val outputFileWithSharpYuv = tempDir.resolve("output_sharp_yuv.webp").toFile()
        val outputFileDefault = tempDir.resolve("output_default.webp").toFile()

        fun WebPWriteParam.prepareParams() {
            compressionType = CompressionType.Lossy
            compressionQuality = 0.95f
        }
        writeWebpImage(
            input = inputImage,
            target = outputFileWithSharpYuv,
            params = {
                prepareParams()
                useSharpYUV = true
            },
        )
        writeWebpImage(
            input = inputImage,
            target = outputFileDefault,
            params = {
                prepareParams()
                useSharpYUV = false
            },
        )

        val outputImageWithYuv = ImageIO.read(outputFileWithSharpYuv)
        val referenceWithYuv = ImageIO.read(getResourceStream("test4_sharp.webp"))
        val outputImageDefault = ImageIO.read(outputFileDefault)
        val referenceDefault = ImageIO.read(getResourceStream("test4_default.webp"))

        assertThat(outputImageDefault).usingComparator(::imagesComparator).isEqualTo(referenceDefault)
        assertThat(outputImageWithYuv).usingComparator(::imagesComparator).isEqualTo(referenceWithYuv)
    }

    private fun imagesComparator(img1: BufferedImage, img2: BufferedImage): Int {
        if (img1.width == img2.width && img1.height == img2.height) {
            for (x in 0 until img1.width) {
                for (y in 0 until img1.height) {
                    if (img1.getRGB(x, y) != img2.getRGB(x, y)) {
                        return img1.getRGB(x, y) - img2.getRGB(x, y)
                    }
                }
            }
            return 0
        } else {
            error("they are not equal at all 🤷‍")
        }
    }
}
